src/languages/fsharp.js

```diff
@@ -84,63 +78,368 @@ export default function(hljs) {
     "yield"
   ];
 
+  const PREPROCESSOR_KEYWORDS = [
+    "if",
+    "else",
+    "endif",
+    "line",
+    "nowarn",
+    "light",
+    "r",
+    "i",
+    "I",
+    "load",
+    "time",
+    "help",
+    "quit"
+  ];
+
+  const LITERALS = [
+    "true",
+    "false",
+    "null",
+    "Some",
+    "None",
+    "Ok",
+    "Error",
+    "infinity",
+    "infinityf",
+    "nan",
+    "nanf",
+    "__LINE__",
+    "__SOURCE_DIRECTORY__",
+    "__SOURCE_FILE__"
```

Comment:
Did you consider `variable.language` or `variable.constant`?  These look like contstants to me, not literals.

Reply:
Oops, I didn't notice these scopes!

I'm not sure I understand all the nuances between `keyword`, `built_in`, `literal`, `variable.language`, and `variable.constant`... ([the doc](https://highlightjs.readthedocs.io/en/latest/css-classes-reference.html) is sometimes confusing, e.g. for `built_in` it mentions it could be used for constants?)

Which one should be used for user-defined constants, and which one for language level constants/macro?
As I understand it,  `__LINE__` and co are some kind of magical constants whose values are set by the compiler. Does that fit with `variable.constant`?

Reply:
> I'm not sure I understand all the nuances between keyword, built_in, literal, variable.language, and variable.constant... 

Me either, LOL... it's a moving target sometimes (with the new ones).  The more nuanced ones are newer scopes and we're trying to use them more.  In the past the older scopes did more "work"...

With regard to current understanding of the older scopes:

- keyword - key language reserved words (if, else, break, continue, etc)
- built_in (built in library functions or variables) - though now this can intermingle with `variable.language` as well as `title.function` and can sometimes be nebulous
- literals - the name is literally the value, `true`, `false`, `null`


> Which one should be used for user-defined constants

`variable.constant`

> which one for language level constants/macro?

I think `variable.constant` as `variable.language` often refers to things such as `this`, `self`, etc...

Reply:
Looks like F# allows you to define your own self identifier so sadly that's not something we can highlight.

Reply:
Ok, thanks for the explanations!

> Looks like F# allows you to define your own self identifier so sadly that's not something we can highlight.

Are you refering to `this` or `self`? If so, I think it's more that F# does not define such a keyword, and allows the developer to use whatever variable name they want to refer to the current object: https://docs.microsoft.com/en-us/dotnet/fsharp/language-reference/classes#self-identifiers